verif/tests/custom/debug_test/debugger.S

```diff
@@ -0,0 +1,643 @@
+	//   15:12 action    = 1
+	//      6  m(achine) = 1
+	li   t1, 2<<28 | 1<<27 | 1<<12 | 1<<6
+	csrr t2,tdata1
+	bne  t1,t2,_debugger_fail
+	csrr t2,tselect
+	bne  x0,t2,_debugger_fail
+	csrr t2,tdata2
+	bne  x0,t2,_debugger_fail
+	csrr t2,tdata3
+	bne  x0,t2,_debugger_fail
+
+	j _debugger_end
+
+_debugger_ebreak:
+    li t0, 4<<28 | 3<<6 | 3<<0
+    csrr t1, dcsr
+    bne t0, t1, _debugger_fail
+        // Increment glb_debug_status
+        la a1, glb_debug_status
+        lw t1, 0(a1)
+        addi t1,t1,1
+        sw t1, 0(a1)
+	// Repeat executing debug code until debug status = hart_status + 3
+	addi t0, t2, 3
+	beq  t1, t0, _debugger_end
+    // Execute non-compressed ebreak for iteration 2
+    addi t0, t2, 2
+    beq t1, t0, _uncompressed_ebreak
+	// Debugger Un-Stack and call debugger code from start using ebreak
+    csrr t0, 0x7b3
+	lw   t1, 4(a0)
+        lw   t2, 8(a0)
+        lw   a1, 12(a0)
+        lw   a2, 16(a0)
+        csrr a0, dscratch
+        ebreak
+_uncompressed_ebreak:
+	// Debugger Un-Stack and call debugger code from start using ebreak
+    csrr t0, 0x7b3
+	lw   t1, 4(a0)
+        lw   t2, 8(a0)
+        lw   a1, 12(a0)
+        lw   a2, 16(a0)
+        csrr a0, dscratch
+        .4byte 0x00100073 # ebreak
+
+_debugger_trigger_in_debug:
+    // setup address to trigger on
+	la   a1, _debugger_trig_point
+	csrw tdata2,a1
+	li   t1, 1<<2
+	csrw tdata1,t1
+	li   t1, 2<<28 | 1<<27 | 1<<12 | 1<<6 | 1 <<2
+	csrr t2,tdata1
+	bne  t1,t2,_debugger_fail
+
+    // Clear glb_expect_debug_entry
+    // If we trig, we'll reenter debug and
+    // test will fail due to 0 flag
+    la a1, glb_expect_debug_entry
+    li t1, 0
+    sw t1, 0(a1)
+_debugger_trig_point:
+    // Should _not_trig here
+    nop
+    // Clear trigger
+    li t1, 0<<2
+    csrw tdata1, t1
+    j _debugger_end
+
+_debugger_trigger_disabled_in_debug:
+    // setup address to trigger on
+	la   a1, _debugger_trig_point_dis
+    // Set trig enable to 0
+	csrw tdata2,a1
+	li   t1, 0<<2
+	csrw tdata1,t1
+	li   t1, 2<<28 | 1<<27 | 1<<12 | 1<<6 | 0 <<2
+	csrr t2,tdata1
+	bne  t1,t2,_debugger_fail
+
+    // Clear glb_expect_debug_entry
+    // If we trig, we'll reenter debug and
+    // test will fail due to 0 flag
+    la a1, glb_expect_debug_entry
+    li t1, 0
+    sw t1, 0(a1)
+_debugger_trig_point_dis:
+    // Should _not_trig here
+    nop
+    // Clear trigger
+    li t1, 0<<2
+    csrw tdata1, t1
+    j _debugger_end
+_debugger_wfi_test:
+    la a1, glb_debug_status
+    csrr t2,dcsr
+    // ebreakm is set by previous test
+    li t1, 4<<28 | 3<<6 | 3<<0 | 1<<15
+    bne t1, t2, _debugger_fail
+
+    // If the following wfi is not converted
+    // to a nop, test will hang
+    wfi
+    j  _debugger_end
+
+_debugger_end:
+    // Check counter values. They should have increased while in debug
+    // regardless of stopcount bit in csr
+    csrr t1, mcycle
+    la a1, glb_mcycle_start
+    lw t2, 0(a1)
+    sub t1, t1, t2
+    beq t1, x0, _debugger_fail
```

Comment:
This part fails on our current CVA6 version .

Why should mcycle increase **regardless** of stopcount bit in csr ? It is stated otherwise in the stopcount bit description in the riscv debug specification (rev. 1.0.0-rc3)

Reply:
Thanks @LQUA , I agree with you on this. It seems been introduced in this commit: https://github.com/openhwgroup/core-v-verif/commit/02016b5dcacd380f6bfb6b598a8aa898a8f01003  

By the way, there might be one issue in current cva6 on this, here:
https://github.com/openhwgroup/cva6/issues/2297